### PR TITLE
Add Notebooks layer freezing

### PR DESCRIPTION
**Pull request by timmy00**
### Goals :soccer:
Add Notebooks  for multi-optimzier and layer freezing

[Slides of final internship presentation](https://nvidia-my.sharepoint.com/:p:/p/mengyaox/ETGXMs5QthdHrFg3kuTmUU4BFYd1kR5I1tnCl60P5OASFg?e=OWeu5V)

### CODE REVIEW

**Review by bschifferer — CHANGES_REQUESTED** [review]
Can you add a unit test for both notebooks similar to https://github.com/NVIDIA-Merlin/models/tree/main/tests/unit/tf/examples

The unittest should use synthetic data

File: examples/08-Accelerate-training-of-larege-embedding-tables-by-LazyAdam.ipynb [hunk 1]
```diff
@@ -0,0 +1,601 @@
+{
```
Comment (bschifferer) [inline examples/08-Accelerate-training-of-larege-embedding-tables-by-LazyAdam.ipynb:1]: <p>spelling mistake <code>nomal</code> ? should be normal? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb [hunk 2]
```diff
@@ -0,0 +1,853 @@
+{
```
Comment (bschifferer) [inline examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb:1]: <p>The headings are not consistent with the previous example.</p><p> </p><p>There the main heading was Generate Synthetic Data</p><p> </p><p>Here you have Building a Two Tower Model ... -&gt; Generate Synthetic Data</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (bschifferer) [inline examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb:1]: <p>spelling mistake</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (bschifferer) [inline examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb:1]: <p>can we reformat it to make it easier to read?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (bschifferer) [inline examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb:1]: <p>we should add the parameter return_dict=True</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (bschifferer) [inline examples/09-Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb:1]: <p>Can we add - Why do we want to use different learning rates for different layers during finetuning?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/667/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

**Comment by bschifferer** [comment]
@timmy00 thanks, that are great examples to add to our library. Thanks for the work. I added a few comments in the notebooks.

1. We need a unit test for each of the notebooks

@rnyak @radekosmulski What do you think where we want to add the notebooks?
I do not think the flow works:
- 01-06 are getting started, etc. notebooks of tensorflow
- 07 is XGB
- 08 is multiple optimizer
- 09 is freezing layers

I think the notebooks are similar to the pre-trained embedding examples.

I think we should create a new folder - something like `advanced-neural-networks-methods` and we add the pretrained + these two examples there (without numbers)?

**Comment by bschifferer** [comment]
We need to update the logos to track them - each notebook requires a unique nvidia_logo.png

**Comment by rnyak** [comment]
> @timmy00 thanks, that are great examples to add to our library. Thanks for the work. I added a few comments in the notebooks.
> 
> 1. We need a unit test for each of the notebooks
> 
> @rnyak @radekosmulski What do you think where we want to add the notebooks? I do not think the flow works:
> 
> * 01-06 are getting started, etc. notebooks of tensorflow
> * 07 is XGB
> * 08 is multiple optimizer
> * 09 is freezing layers
> 
> I think the notebooks are similar to the pre-trained embedding examples.
> 
> I think we should create a new folder - something like `advanced-neural-networks-methods` and we add the pretrained + these two examples there (without numbers)?

I like the idea of creating a new folder `advanced-neural-networks-methods` and adding examples without numbers.

**Comment by radekosmulski** [comment]
There is a typo in the first notebook, nexted -> nested, I believe you can run `pre-commit install` in the repo and it will run automated checks for you before committing, I find this quite helpful.

In the second notebook, I had to read through all of this to understand the value on freezing layers. It might be worth considering whether adding more prose on top of the notebook wouldn't be helpful in justifying why we might want to go for this method.

I am with @rnyak that moving these notebooks to a subdirectory might be a good idea. I would however opt for a shorter name, something like `advanced_methods`, who knows what we might want to put there down the road 🙂

File: examples/08-Accelerate-training-of-larege-embedding-tables-by-LazyAdam.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (rnyak) [reply examples/08-Accelerate-training-of-larege-embedding-tables-by-LazyAdam.ipynb:1]: <p>agreed that last part does not fit into this notebook. I will remove it. We can add W&amp;D model into notebook 3 but Gabriel already created a separate example nb for w&amp;d didnt he? </p>

Comment (rnyak) [reply examples/08-Accelerate-training-of-larege-embedding-tables-by-LazyAdam.ipynb:1]: replaced `normal` with `original`.

**Comment by rnyak** [comment]
The second notebook `Incremental-Training-with-Different-Learning-Rates-and-Layer-Freezing.ipynb` is broken, it gives error.

**Comment by bschifferer** [comment]
The multi-optimizer use case was created as a new PR: https://github.com/NVIDIA-Merlin/models/pull/747

**Pull request closed without merging**